### DISCUSSION TITLE
Possible to show custom meters on Dashboard Metrics?

### DISCUSSION THREAD

**Discussion by alex-todorov-j2** [microsoft/aspire#2383] [Q&A] [answered]
I added a custom meter thinking that it would show up on the Metrics page of the Dashboard, but I still see only the defaults:

Microsoft.AspNetCore.Hosting
Microsoft.AspNetCore.Server.Kestrel
OpenTelemetry.Instrumentation.Runtime
System.Net.Http

Wasn't the idea of the Dashboard combined with the OTLP exporter to show custom meters too? 

Otherwise you have to use a specific exporter like Prometheus and visualize custom metrics in a separate tool like Grafana.

**Comment by davidfowl** [comment, marked as answer]
You need to explicitly add code to your application to make it show up.

```C#
builder.Services.AddOpenTelemetry().WithMetrics(m => m.AddMeter("YourMeter"));
```

**Reply by alex-todorov-j2** [reply]
Yes, that worked. Thank you David! I was adding it wrongly with: `meterProviderBuilder.AddMeter("YourMeter");`

One question: should custom meters be registered in AspireStarterApp.ServiceDefaults or in specific services, like AspireStarterApp.ApiService in this case? Is AspireStarterApp.ServiceDefaults purely for .NET Aspire built-in services?

**Reply by davidfowl** [reply]
Service defaults if it applies to all services

**Comment by FMauroy** [comment]
Hi, how can we achieve the same results when using the Aspire Dashboard as an independent Docker container?
My customer meter is configured as per your suggestion but it doesn't appear in the dashboard...
Thanks!

**Reply by davidfowl** [reply]
The exact same thing I described should work in either scenario. Do you have the standalone dashboard working with your application? 

Good news, the docs are up to date https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/standalone

**Reply by FMauroy** [reply]
Thanks for the quick reply.

Yes, I was referring to this documentation, maybe I missed a step.

Cheers
fred

**Reply by davidfowl** [reply]
Can you share a repository that shows the problem? Here's the sample we have linked from the documentation https://github.com/dotnet/aspire-samples/tree/main/samples/StandaloneDashboard

**Reply by FMauroy** [reply]
I'm able to see the structured logs and traces, but for metrics, I only get default ones:

![image](https://github.com/dotnet/aspire/assets/1107417/b1c9d1b7-6d29-4925-9c4f-b1a44ae2f65c)

I expected to see my own metrics, as well as those from Microsoft.AspNetCore.Hosting.

I can see them using dotnet-counters, maybe this is not supported by the Aspire dashboard?

**Reply by davidfowl** [reply]
It is, but I need to see the code you are using to use your own metrics. That's why I want to see a repository with all of the code. It's likely you didn't add a source to the metrics setup.

**Reply by AHansen-Planview** [reply]
There does appear to be some undocumented restrictions on allowed characters for naming of measurements. 

Best practices states:
_.NET doesn't enforce any naming scheme for Instruments, but we recommend following [OpenTelemetry naming guidelines](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/metrics.md#general-guidelines), which use lowercase dotted hierarchical names and an underscore ('_') as the separator between multiple words in the same element._

However, using a measurement name that mimics a generic (i.e.` Handler<GetUserByIdQuery>.count`), none of the exporters appear to output that data. Meanwhile, Dotnet-counters show the metric, measurements and counts as expected. Altering the name to not include <> solves the omissions. This appears to be applied somewhere within the OpenTelemetry packages, and not within Aspire itself. 

Apologies for a non-thorough investigation as to where this is happening, time is always a limitation. Hoped this would be helpful to get it out there incase others are having a similar experience.
